pkg/cloud/external_storage.go

```diff
@@ -183,3 +183,43 @@ type ExternalStorageOptions struct {
 type ExternalStorageConstructor func(
 	context.Context, ExternalStorageContext, cloudpb.ExternalStorage,
 ) (ExternalStorage, error)
+
+// NewExternalStorageAccessor creates an uninitialized ExternalStorageAccessor.
+func NewExternalStorageAccessor() *ExternalStorageAccessor {
+	return &ExternalStorageAccessor{ready: make(chan struct{})}
+}
+
+// ExternalStorageAccessor is a container for accessing the ExternalStorage
+// factory methods once they are initialized. Attempts to access them prior to
+// initialization will block.
+type ExternalStorageAccessor struct {
+	ready   chan struct{}
+	factory ExternalStorageFactory
+	byURI   ExternalStorageFromURIFactory
+}
+
+// Init initializes the ExternalStorageAccessor with the passed factories.
+func (a *ExternalStorageAccessor) Init(
+	factory ExternalStorageFactory, uriFactory ExternalStorageFromURIFactory,
+) error {
+	close(a.ready)
+	a.factory = factory
```

Comment:
we should move this _above_ the close right? in case the two happen concurrently. That way by the time the channel is closed, `a.factory` is set.